### PR TITLE
[jit] split canonicalize_ops, make a decompose pass

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20039 dispatch and expose linear op
* **#19988 split canonicalize_ops, make a decompose pass**
* #19987 lower batchmm to non-diff optimization

Summary:
This PR enables the `addmm` decomposition happens later (after custom fusion, before batchmm), this will enable the custom fusion to see high level operator informations (such as `addmm` and `linear`, instead of `mm` + `add`)

Test plan:
Test if the decomposition pass works. 

Differential Revision: [D15190355](https://our.internmc.facebook.com/intern/diff/D15190355)

### CODE REVIEW

File: torch/csrc/jit/graph_executor.cpp [hunk 1]
```diff
@@ -645,6 +646,10 @@ struct GraphExecutorImpl {
     }
     // decomposition pass, decompose certain ops that will be used in the following
     // passes (like batchmm and jit fusion)
+    DecomposeOps(graph);
+    // Autodiff and decompositon pass will replace some parts of graph with new graph
+    // these new graphs usually miss shape information on nodes, so we propagate shapes
+    // PropagateInputShapes(graph);
```
Comment (bwasti) [inline torch/csrc/jit/graph_executor.cpp]: why is this commented out?

**Review by bwasti — APPROVED** [review]

File: torch/csrc/jit/graph_executor.cpp (hunk 1 above)
Comment (wanchaol) [reply torch/csrc/jit/graph_executor.cpp]: some issues on `nn.Dropout` when enable this, I will need to look into that and will bring the pass back.

File: torch/csrc/jit/passes/decompose_ops.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+#include <torch/csrc/jit/passes/decompose_ops.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/passes/constant_propagation.h>
+#include <torch/csrc/jit/passes/shape_analysis.h>
+#include <torch/csrc/jit/passes/utils/subgraph_utils.h>
+
+namespace torch {
+namespace jit {
+
+Value* decomposeOp(
```
Comment (zdevito) [inline torch/csrc/jit/passes/decompose_ops.cpp]: What is this suppose to do? If it is suppose to execute this once, it doesn't work, it is allocating a std::once_flag each time through the function.

refactor:
```
static std::shared_ptr<Graph> addmm_graph = createAddMMGraph();
```

Comment (zdevito) [reply torch/csrc/jit/passes/decompose_ops.cpp]: I would also prefer if we added a constructor to compilation unit so this can just be write. I don't want it to end up being the case that everyone copy/pastes this pattern.

static script::CompilationUnit addmm("....");

**Review by zdevito — COMMENTED** [review]
I think the pass changes are good. Can you address the comments about using compilation unit to load code?

File: torch/csrc/jit/graph_executor.cpp [hunk 3]
```diff
@@ -638,11 +640,26 @@ struct GraphExecutorImpl {
     CheckInplace(graph);
   }
 
+  void runPostdiffOptimization(std::shared_ptr<Graph>& graph){
+    // Run post differentiation optimization passes, Autodiff will replace some parts of graph with new graph 
+    // these new graphs usually consists of control flow and miss shape information on nodes, so we run shape
+    // prop and constant prop, as well as loop unrolling in some cases
+    PropagateInputShapes(graph);
+    ConstantPropagation(graph);
```
Comment (eellison) [inline torch/csrc/jit/graph_executor.cpp]: This should call peephole optimization as well i think.

Comment (wanchaol) [reply torch/csrc/jit/graph_executor.cpp]: `runOptimization` have bunch of other passes that we don't need to run, just want to minimize the passes that needs to be run, plus it does not run `propagateInputShapes`, so I just make a separate group of passes.

Comment (eellison) [reply torch/csrc/jit/graph_executor.cpp]: Could you call `propagateInputShapes` and then call into `runOptimization`? The downside is that we're now maintaining two separate sets of optimizations, and you might miss optimizations that you think you didn't need (as with prim::is_cuda because PeepholeOptimize wasn't called)

Comment (wanchaol) [reply torch/csrc/jit/graph_executor.cpp]: sounds legit. the only concern i was having is that this will increase our compilation time quite a bit, that's why I want separate the passes in the first place. But given that we are potentially missing bunch of optimizations, I will change it to just run the optimization :)
